python/demo/demo_scipy-solvers.py

```diff
@@ -0,0 +1,146 @@
+# ---
+# jupyter:
+#   jupytext:
+#     text_representation:
+#       extension: .py
+#       format_name: light
+#       format_version: '1.5'
+#       jupytext_version: 1.13.6
+# ---
+# Copyright (C) 2024 Jørgen S. Dokken
+#
+# This file is part of DOLFINx (<https://www.fenicsproject.org>)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+#
+# # $L^2$ projection through minimization using SciPy solvers
+# In this demo, we will go through how to do a `projection` of a spatially varying function `g`,
+# into a function space `V`.
+#
+# An $L^2$ projection can be viewed as the following minimization problem:
+#
+# $$ \min_{u\in V} G(u) = \frac{1}{2}\int_\Omega (u-g)\cdot (u-g)~\mathrm{d}x.$$
+
+# We start by importing the necessary modules
+
+from mpi4py import MPI
+
+import numpy as np
+import numpy.typing as npt
+import scipy.optimize
+import scipy.sparse.linalg
+
+import dolfinx
+import ufl
+
+if MPI.COMM_WORLD.size > 1:
+    print("This demo works only in serial.")
+    exit(0)
+
+# Next we define the `domain` ($\Omega$) and the function space `V` where we want to
+# project the function `g` into.
+
+Nx = 15
+Ny = 20
+domain = dolfinx.mesh.create_unit_square(MPI.COMM_WORLD, Nx, Ny)
+V = dolfinx.fem.functionspace(domain, ("Lagrange", 1))
+x, y = ufl.SpatialCoordinate(domain)
+g = ufl.sin(ufl.pi * x) * ufl.cos(ufl.pi * y)
+
+# The functional $G$ can be defined as follows
+
+uh = dolfinx.fem.Function(V)
+G = ufl.dot(uh - g, uh - g) * ufl.dx
+
+G_compiled = dolfinx.fem.form(G)
+
+# We can find the minimum of this function by differentiating
+# the function with respect to `u`:
+#
+# $$ F(u, \delta u) = \frac{\mathrm{d}G}{\mathrm{d}u}[\delta u] = 0, \quad \forall \delta u \in V.$$
+
+du = ufl.conj(ufl.TestFunction(V))
+residual = ufl.derivative(G, uh, du)
+
+# We generate the integration kernels for the residual, and assemble an initial
+# residual vector `b`.
+
+F = dolfinx.fem.form(residual)
+b = dolfinx.fem.assemble_vector(F)
+
+# A convenience function for computing the value of the functional
+
+
+def compute_functional_value() -> dolfinx.default_scalar_type:
+    local_contribution = dolfinx.fem.assemble_scalar(G_compiled)
+    return domain.comm.allreduce(local_contribution, op=MPI.SUM)
+
+
+# We can use the scipy Newton-Krylov solver to solve this problem.
+# This solver takes in a function that computes the residual of the functional
+# We re-use our structures for `uh` and `b` to avoid unnecessary allocations.
+
+
+def compute_residual(x) -> npt.NDArray[dolfinx.default_scalar_type]:
+    """
+    Evaluate the residual F(x) = 0
+
+    :param x: Input vector with current solution
+    :return: Returns residual array
```

Comment:
Doesn't FEniCS use the Google style for docs? Also does the docstring really matter for a function in a demo?
```suggestion
    Args:
        x: Input vector with current solution
    Returns:
        Residual array
```

Reply:
It’s good practice to document I guess.